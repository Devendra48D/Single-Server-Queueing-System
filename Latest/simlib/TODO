
TODO list for SIMLIB/C++
========================

x86_64 128bit "red zone"

= update INSTALL/README

- IntegratorLim implementation
- simplify numerical methods framework
- autodetect multilib and create 32+64bit static+shared libs
- use autotools?

DONE - run ERRMSG generator only if really needed (timestamp-based)

- try clang++

- add soname and symlinks, OR use libtool?

- add check for version vs name of tarball

- add more Doxygen documentation to source code

- automate SIMLIB script generation by make install

- create both 32 and 64 bit versions, use suffix -32, -64, and symlink ?

- add reference counting to aContiBlock  (+rename) OR use shared_ptr<>

- tests for performance (regression tests, calendar)

porting: windows/mingw, windows/vc

use generic settings in CXX
version 2.90...3.0

- New monitoring (NOT function = Evaluate each step) block: 
  TotalMinMax(x) + methods GetMin/GetMax/GetAvg/effective?
  LastPeakValue +-
  Integrator: add Total min/max
+ block+variable RelativeSimulationTime  0=begin 1=endtime
     works only if endtime given - else error|0 ?

- improve code - see src/TODO
- add configure script
- testing...

- Init(...) should INITIALIZE (Clear) all time statistics


======================================================
- new debuging:

DebugON()               Debug(1)
DebugOFF()              Debug(0)

Debug(0|1|level);

levels:
0 - none
1 - basic operation
2 - more
4 - detail
5 - all
======================================================

