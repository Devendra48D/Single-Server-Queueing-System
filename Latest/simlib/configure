#!/bin/sh

echo configure
echo TODO: generate Makefiles
echo For now try make, make 32, make 64, make clean-all, ... 
echo Sorry.
