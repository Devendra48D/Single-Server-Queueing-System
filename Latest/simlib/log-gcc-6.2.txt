make -C src
make[1]: Entering directory '/disk2/peringer/WORK/SIMLIB/simlib/src'
make -j3 -f Makefile.`uname -s`-`uname -m`
make[2]: Entering directory '/disk2/peringer/WORK/SIMLIB/simlib/src'
cc     generr.c   -o generr
./generr errors.txt
./generr errors.txt
GENERR - error messages generator                     (c)PerPet 1991-1997 

Number of messages = 89 
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c atexit.cc
GENERR - error messages generator                     (c)PerPet 1991-1997 

Number of messages = 89 
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c calendar.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c debug.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c entity.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c error.cc
entity.cc: In member function ‘virtual void simlib3::Entity::Activate(double)’:
entity.cc:27:24: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKENTITY(e) if (!e) SIMLIB_error(EntityRefError);
                        ^
entity.cc:63:3: note: in expansion of macro ‘CHECKENTITY’
   CHECKENTITY(this);
   ^~~~~~~~~~~
entity.cc: In member function ‘virtual void simlib3::Entity::Passivate()’:
entity.cc:27:24: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKENTITY(e) if (!e) SIMLIB_error(EntityRefError);
                        ^
entity.cc:86:3: note: in expansion of macro ‘CHECKENTITY’
   CHECKENTITY(this);
   ^~~~~~~~~~~
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c errors.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c event.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c link.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c list.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c name.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c object.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c print.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c run.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c sampler.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c opt-hooke.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c opt-simann.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c opt-param.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c delay.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c zdelay.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c simlib2D.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c simlib3D.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c algloop.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c cond.cc
algloop.cc: In member function ‘virtual double simlib3::Bisect::Value()’:
algloop.cc:182:10: warning: variable ‘fa’ set but not used [-Wunused-but-set-variable]
   double fa = 0; // function value in boundary point
          ^~
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c fun.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c graph.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c intg.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c continuous.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c ni_abm4.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c ni_euler.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c ni_fw.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c ni_rke.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c ni_rkf3.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c ni_rkf5.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c ni_rkf8.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c numint.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c output1.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c stdblock.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c barrier.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c facility.cc
facility.cc: In member function ‘virtual void simlib3::Facility::QueueIn(simlib3::Entity*, simlib3::ServicePriority_t)’:
facility.cc:33:29: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKFACILITY(fptr) if (!fptr) SIMLIB_error(FacilityRefError)
                             ^
facility.cc:210:5: note: in expansion of macro ‘CHECKFACILITY’
     CHECKFACILITY(this);
     ^~~~~~~~~~~~~
facility.cc: In member function ‘virtual void simlib3::Facility::Release(simlib3::Entity*)’:
facility.cc:33:29: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKFACILITY(fptr) if (!fptr) SIMLIB_error(FacilityRefError)
                             ^
facility.cc:160:5: note: in expansion of macro ‘CHECKFACILITY’
     CHECKFACILITY(this);
     ^~~~~~~~~~~~~
facility.cc: In member function ‘virtual void simlib3::Facility::Seize(simlib3::Entity*, simlib3::ServicePriority_t)’:
facility.cc:33:29: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKFACILITY(fptr) if (!fptr) SIMLIB_error(FacilityRefError)
                             ^
facility.cc:119:5: note: in expansion of macro ‘CHECKFACILITY’
     CHECKFACILITY(this);
     ^~~~~~~~~~~~~
facility.cc: In member function ‘void simlib3::Facility::SetQueue(simlib3::Queue*)’:
facility.cc:33:29: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKFACILITY(fptr) if (!fptr) SIMLIB_error(FacilityRefError)
                             ^
facility.cc:98:5: note: in expansion of macro ‘CHECKFACILITY’
     CHECKFACILITY(this);
     ^~~~~~~~~~~~~
facility.cc: In member function ‘virtual void simlib3::Facility::Clear()’:
facility.cc:33:29: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKFACILITY(fptr) if (!fptr) SIMLIB_error(FacilityRefError)
                             ^
facility.cc:273:5: note: in expansion of macro ‘CHECKFACILITY’
     CHECKFACILITY(this);
     ^~~~~~~~~~~~~
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c histo.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c output2.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c process.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c queue.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c random1.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c random2.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c semaphor.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c stat.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c store.cc
stat.cc: In member function ‘void simlib3::Stat::operator()(double)’:
stat.cc:36:3: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
   if (!this)  SIMLIB_error(StatRefError);
   ^~
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c tstat.cc
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c waitunti.cc
store.cc: In member function ‘virtual void simlib3::Store::QueueIn(simlib3::Entity*, long unsigned int)’:
store.cc:34:26: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKSTORE(sptr) if (!sptr) SIMLIB_error(StoreRefError)
                          ^
store.cc:187:3: note: in expansion of macro ‘CHECKSTORE’
   CHECKSTORE(this);
   ^~~~~~~~~~
store.cc: In member function ‘virtual void simlib3::Store::Enter(simlib3::Entity*, long unsigned int)’:
store.cc:34:26: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKSTORE(sptr) if (!sptr) SIMLIB_error(StoreRefError)
                          ^
store.cc:138:3: note: in expansion of macro ‘CHECKSTORE’
   CHECKSTORE(this);
   ^~~~~~~~~~
store.cc: In member function ‘virtual void simlib3::Store::Leave(long unsigned int)’:
store.cc:34:26: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKSTORE(sptr) if (!sptr) SIMLIB_error(StoreRefError)
                          ^
store.cc:161:3: note: in expansion of macro ‘CHECKSTORE’
   CHECKSTORE(this);
   ^~~~~~~~~~
store.cc: In member function ‘void simlib3::Store::SetCapacity(long unsigned int)’:
store.cc:34:26: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKSTORE(sptr) if (!sptr) SIMLIB_error(StoreRefError)
                          ^
store.cc:106:3: note: in expansion of macro ‘CHECKSTORE’
   CHECKSTORE(this);
   ^~~~~~~~~~
store.cc: In member function ‘void simlib3::Store::SetQueue(simlib3::Queue*)’:
store.cc:34:26: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKSTORE(sptr) if (!sptr) SIMLIB_error(StoreRefError)
                          ^
store.cc:118:3: note: in expansion of macro ‘CHECKSTORE’
   CHECKSTORE(this);
   ^~~~~~~~~~
store.cc: In member function ‘virtual void simlib3::Store::Clear()’:
store.cc:34:26: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKSTORE(sptr) if (!sptr) SIMLIB_error(StoreRefError)
                          ^
store.cc:198:3: note: in expansion of macro ‘CHECKSTORE’
   CHECKSTORE(this);
   ^~~~~~~~~~
tstat.cc: In member function ‘virtual void simlib3::TStat::operator()(double)’:
tstat.cc:29:28: warning: nonnull argument ‘this’ compared to NULL [-Wnonnull-compare]
 #define CHECKTSTAT(tsptr)  if (!tsptr)  SIMLIB_error(TStatRefError);
                            ^
tstat.cc:70:3: note: in expansion of macro ‘CHECKTSTAT’
   CHECKTSTAT(this);
   ^~~~~~~~~~
g++ -Wall -fPIC   -m64         -O2          -g           -Wextra      -c version.cc
rm -f simlib.a  # create new library
g++ -shared -Wall -fPIC   -m64         -O2          -g           -Wextra      version.o atexit.o calendar.o debug.o entity.o error.o errors.o event.o link.o list.o name.o object.o print.o run.o sampler.o opt-hooke.o opt-simann.o opt-param.o delay.o zdelay.o simlib2D.o simlib3D.o algloop.o cond.o fun.o graph.o intg.o continuous.o ni_abm4.o ni_euler.o ni_fw.o ni_rke.o ni_rkf3.o ni_rkf5.o ni_rkf8.o numint.o output1.o stdblock.o barrier.o facility.o histo.o output2.o process.o queue.o random1.o random2.o semaphor.o stat.o store.o tstat.o waitunti.o -o simlib.so
ar rcv simlib.a atexit.o calendar.o debug.o entity.o error.o errors.o event.o link.o list.o name.o object.o print.o run.o sampler.o opt-hooke.o opt-simann.o opt-param.o delay.o zdelay.o simlib2D.o simlib3D.o algloop.o cond.o fun.o graph.o intg.o continuous.o ni_abm4.o ni_euler.o ni_fw.o ni_rke.o ni_rkf3.o ni_rkf5.o ni_rkf8.o numint.o output1.o stdblock.o barrier.o facility.o histo.o output2.o process.o queue.o random1.o random2.o semaphor.o stat.o store.o tstat.o waitunti.o version.o
a - atexit.o
a - calendar.o
a - debug.o
a - entity.o
a - error.o
a - errors.o
a - event.o
a - link.o
a - list.o
a - name.o
a - object.o
a - print.o
a - run.o
a - sampler.o
a - opt-hooke.o
a - opt-simann.o
a - opt-param.o
a - delay.o
a - zdelay.o
a - simlib2D.o
a - simlib3D.o
a - algloop.o
a - cond.o
a - fun.o
a - graph.o
a - intg.o
a - continuous.o
a - ni_abm4.o
a - ni_euler.o
a - ni_fw.o
a - ni_rke.o
a - ni_rkf3.o
a - ni_rkf5.o
a - ni_rkf8.o
a - numint.o
a - output1.o
a - stdblock.o
a - barrier.o
a - facility.o
a - histo.o
a - output2.o
a - process.o
a - queue.o
a - random1.o
a - random2.o
a - semaphor.o
a - stat.o
a - store.o
a - tstat.o
a - waitunti.o
a - version.o
ranlib simlib.a
nm --demangle simlib.a > simlib.nm
make[2]: Leaving directory '/disk2/peringer/WORK/SIMLIB/simlib/src'
make[1]: Leaving directory '/disk2/peringer/WORK/SIMLIB/simlib/src'
make -C examples
make[1]: Entering directory '/disk2/peringer/WORK/SIMLIB/simlib/examples'
make -j3 -f Makefile.`uname -s`-`uname -m`
make[2]: Entering directory '/disk2/peringer/WORK/SIMLIB/simlib/examples'
g++ -Wall -m64 -O2  -g   -I../src -o bessel  bessel.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o centrala  centrala.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o ctest  ctest.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o kolo  kolo.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o kolo2  kolo2.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o kostky  kostky.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o kostky2  kostky2.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o lorenz  lorenz.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o micek  micek.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o ball2  ball2.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o model2  model2.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o model3  model3.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o model4  model4.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o model5  model5.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o multiexp  multiexp.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o pocsit  pocsit.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o procesor  procesor.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test3D  test3D.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o topeni  topeni.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o model2-timeout  model2-timeout.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o rc  rc.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o rlc  rlc.cc ../src/simlib.so -lm
make[2]: Leaving directory '/disk2/peringer/WORK/SIMLIB/simlib/examples'
make[1]: Leaving directory '/disk2/peringer/WORK/SIMLIB/simlib/examples'
make -C tests
make[1]: Entering directory '/disk2/peringer/WORK/SIMLIB/simlib/tests'
make -j3 -f Makefile.`uname -s`-`uname -m`
make[2]: Entering directory '/disk2/peringer/WORK/SIMLIB/simlib/tests'
g++ -Wall -m64 -O2  -g   -I../src -o 3d-test  3d-test.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o barrier-test1  barrier-test1.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o barrier-test2  barrier-test2.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o delay-test  delay-test.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o delay-test2  delay-test2.cc ../src/simlib.so -lm
delay-test.cc: In function ‘void Sample()’:
delay-test.cc:28:10: warning: variable ‘error’ set but not used [-Wunused-but-set-variable]
   double error = delayed - e.Value();
          ^~~~~
g++ -Wall -m64 -O2  -g   -I../src -o zdelay-test  zdelay-test.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o waituntil-test  waituntil-test.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o process-test  process-test.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o sizeof-all  sizeof-all.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o random-test  random-test.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test1  test1.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test2  test2.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test3  test3.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test4  test4.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test5  test5.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test-calendar  test-calendar.cc ../src/simlib.so -lm
g++ -Wall -m64 -O2  -g   -I../src -o test-reactivate  test-reactivate.cc ../src/simlib.so -lm
make[2]: Leaving directory '/disk2/peringer/WORK/SIMLIB/simlib/tests'
make[1]: Leaving directory '/disk2/peringer/WORK/SIMLIB/simlib/tests'
