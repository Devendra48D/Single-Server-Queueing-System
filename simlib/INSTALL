
----------------------------------------------------------------
Installation instructions for SimLib/C++ (Simulation Library for C++)
----------------------------------------------------------------

This is SIMLIB3 experimental pre-release version
Warning: Compilation method will change in future
Last modified: 2011-04-12

----------------------------------------------------------------
How to build and install SIMLIB/C++
----------------------------------------------------------------

0. Unpack the source archive.

1. Type "make" or "gmake" (use GNU make)
   If it does not work, check/create Makefile.<OS>-<CPU>
   (On some platforms you should rewrite part of process.cc)
   You can use it directly:   make -f Makefile.<OS>-<CPU>
   You can compile 32bit version on 64bit platform:  make 32

2. Test the library by "make test"  (pointer differences in output are O.K.).

3. Test the library by running examples in ../examples (make run).

4. Type "make install" (installs to /usr/local directory).
   You should set dynamic linker configuration for /usr/local/lib.
   The installation goes to /usr/local/lib, if you want .../lib64 on 64bit
   system, change the makefile.


This version of SIMLIB has been compiled and tested on:

  System                         Compiler
  ------------------------------------------------
  Linux 2.6.26, Athlon64/64bit   (gcc-4.3.2)
  Linux 2.6.26, Athlon64/64bit   (gcc-4.5.0)
  Linux 2.6.32, Core2/64bit      (gcc-4.6.0)

Old versions tested on:
  Linux 2.6.24, Core2duo/32bit   (gcc-4.3.0)
  Linux 2.6.18, Athlon64/64bit   (gcc-4.1.2)
  Linux 2.6.18, Athlon64/32bit   (gcc-4.1.2)
  Linux 2.6.9,  Athlon64/32bit   (gcc-3.4)
  Linux 2.2.18, Pentium 3        (gcc-3.0)
  Linux 2.2.17, Pentium 3        (gcc-2.95.2)
  Linux 2.0.33, i386+            (egcs-1.0.3)
  Linux 2.2.12, Pentium          (egcs-2.91.66)
  FreeBSD 2.2.7-STABLE, i386+    (egcs-2.91.57)
  FreeBSD 3.3-STABLE, Pentium 2  (gcc-2.95.1)
  SCO UNIX_SV 4.2MP 2.1.2, i386+ (egcs-2.91.57)
  MSDOS 5.0+, i386+              (DJGPP=GCC 2.8+)
  ------------------------------------------------

Shared library have been tested on Linux.

Porting to another system is easy, except the process switching (see
process.cc for details) which depends on processor type and stack layout.
If you disable processes, it should work on any system with
(sub)standard C++ compiler.
Patches for other architectures are welcome ;-)

Bug reports please mail to:  peringer AT fit.vutbr.cz

----------------------------------------------------------------
More information is at the SIMLIB home page:

   http://www.fit.vutbr.cz/~peringer/SIMLIB/

----------------------------------------------------------------
Written by Petr Peringer              peringer AT fit.vutbr.cz
