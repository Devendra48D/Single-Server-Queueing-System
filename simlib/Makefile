# Makefile for SIMLIB (unix)

# TODO: fuzzy extension

.PHONY: doc all clean clean-all test clean-doc pack

all:
	make -C src
	make -C examples
	make -C tests

rebuild-all: clean-all
	make all

32:
	make -C src        32
	make -C examples   32
	make -C tests      32
64:
	make -C src        64
	make -C examples   64
	make -C tests      64

test:
	make -C src        test
test32:
	make -C src        test32

###untested
#fuzzy:
#	make -C src fuzzy

install:
	make -C src install
uninstall:
	make -C src uninstall
# cleaning etc.
clean:
	make -C src clean
	make -C examples clean
	make -C tests clean

### quick test
#test:
#	make -C src test

# doxygen-generated documentation
doc:
	doxygen doxygen.config

clean-doc:
	rm -rf ./doc/doxygen
	make -C ./doc/latex clean

clean-all: clean clean-doc
	make -C examples clean-all
	make -C ./doc/latex clean-all
	make -C tests clean-all

# create archive with .git
pack: clean-all
	(N=`basename $$(pwd)`; cd ..; \
         tar czvf $$N-git-`date +"%Y%m%d"`.tar.gz $$N ; \
         echo "Done.")

# create distribution archive
dist: clean-all
	git log >git-log.txt
	touch . # the main directory should have current time stamp
	(N=`basename $$(pwd)`; cd ..; \
         tar czvf $$N-`date +"%Y%m%d"`.tar.gz --exclude-vcs $$N; \
         echo "Done.")


