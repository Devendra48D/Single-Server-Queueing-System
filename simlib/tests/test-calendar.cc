////////////////////////////////////////////////////////////////////////////
// SIMLIB/C++ -- basic calendar test
//
// simple check: increasing time ordering of event execution
//
#include "simlib.h"
#include <cstdlib>
#include <cmath>

long     N         = 1000;
unsigned number    = 0;
double   Last_Time = 0.0;

class TestEvent : public Event {
  void Behavior() {
      if (Time < Last_Time)
          Error("Bad calendar implementation %g < %g", Time, Last_Time);
      Last_Time = Time;
  }
  public:
  TestEvent() { ++number; }
  ~TestEvent() { --number; }
};

// various test distributions of time intervals
// all should have mean==1
double DistributionExp() {
    double x = Exponential(1);
    return x;
}
double DistributionUni() {
    double x = 2*Random();
    return x;
}
double DistributionBiased() {
    double x = 0.9+0.2*Random();
    return x;
}
double DistributionBimodal() {
    double x = 0.95238*Random();
    if(Random()<0.1) x+=9.5238;
    return x;
}
double DistributionTriangular() {
    double x = 1.5*std::sqrt(Random());
    return x;
}

// 
const int N_D = 5;
double (*distribution[N_D])() = {
    DistributionUni,
    DistributionExp,
    DistributionBiased,
    DistributionBimodal,
    DistributionTriangular,
};

// 
double MAX=0;
int dd = 0;
double Distribution() {
    double x = distribution[dd]();
    if (x>MAX) MAX=x;
    return x;
}

// experiment
int main(int argc, char *argv[])
{
    //DebugON();
#if 1
    SetCalendar("cq"); // calendar queue
#endif
    if (argc > 1) {
        N = 0;
        N = std::strtoul(argv[1], 0, 10);
        if (N < 1)
            _Print("Error: bad argument\n");
    }
    Print("Calendar size: %ld\n", N);
    int i = 0;
    for(dd=0; dd < N_D; ++dd) try {
        Init(0);          // Initialize time, calendar, ...
        Last_Time=0.0;
        MAX=0;
        Print("Run#%d ", dd);
        // create and activate N events 
        // we need to:
        //  - preallocate Events
        //  - preallocate EventNotices
        for (i = 0; i < N; i++) {
            (new TestEvent)->Activate(Distribution());
        }
        // Calendar filled
        //Print("max=%g\n", MAX);
        // and now dequeue all
        Run();                  // simulation
        // delete events automatically by end of Behavior()
        Print("OK\n");
    }
    catch(std::bad_alloc) {
        Print("No memory after %d created events\n", i);
    }
    catch(...) {
        Print("Exception\n");
    }
}
