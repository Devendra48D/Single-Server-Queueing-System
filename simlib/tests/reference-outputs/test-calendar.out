Calendar size: 1000
Run#0 OK
Run#1 OK
Run#2 OK
Run#3 OK
Run#4 OK
